File: impls/rust1_94_0/src/bin/step2_eval.rs
use std::collections::HashMap;

use rust1_94_0::{
    error::MalError,
    printer::pr_str,
    reader::read_str,
    readline as rl,
    types::{
        MalFunction, MalHashmap, MalList, MalNumber, MalObject, MalProcedure, MalResult, MalSymbol,
        MalVector, mal_func,
    },
};

type Env = HashMap<String, MalFunction>;

fn repl_env() -> HashMap<String, MalFunction> {
    let mut env = HashMap::new();
    env.insert(
        "+".to_string(),
        mal_func(|args: &[MalObject]| match args {
            &[
                MalObject::Number(MalNumber { n: a, .. }),
                MalObject::Number(MalNumber { n: b, .. }),
            ] => Ok(MalObject::Number(MalNumber::new(a + b))),
            _ => Err(MalError::InvalidArguments),
        }),
    );
    env.insert(
        "-".to_string(),
        mal_func(|args: &[MalObject]| match args {
            &[
                MalObject::Number(MalNumber { n: a, .. }),
                MalObject::Number(MalNumber { n: b, .. }),
            ] => Ok(MalObject::Number(MalNumber::new(a - b))),
            _ => Err(MalError::InvalidArguments),
        }),
    );
    env.insert(
        "*".to_string(),
        mal_func(|args: &[MalObject]| match args {
            &[
                MalObject::Number(MalNumber { n: a, .. }),
                MalObject::Number(MalNumber { n: b, .. }),
            ] => Ok(MalObject::Number(MalNumber::new(a * b))),
            _ => Err(MalError::InvalidArguments),
        }),
    );
    env.insert(
        "/".to_string(),
        mal_func(|args: &[MalObject]| match args {
            &[
                MalObject::Number(MalNumber { n: a, .. }),
                MalObject::Number(MalNumber { n: b, .. }),
            ] => Ok(MalObject::Number(MalNumber::new(a / b))),
            _ => Err(MalError::InvalidArguments),
        }),
    );
    env
}

fn read(input: &str) -> MalResult {
    read_str(input)
}

fn eval(ast: &MalObject, env: &Env) -> MalResult {
    match ast {
        MalObject::Symbol(MalSymbol { name, .. }) => {
            let key = name.clone().to_string();
            if env.contains_key(&key) {
                Ok(MalObject::Procedure(MalProcedure::new(
                    env.get(&key).unwrap().clone(),
                )))
            } else {
                Err(MalError::RuntimeError(format!("'{key}' not found")))
            }
        }
        MalObject::Vector(MalVector { items, .. }) => items
            .iter()
            .try_fold(Vec::new(), |mut acc, arg| {
                let item = eval(arg, env)?;
                acc.push(item);
                Ok(acc)
            })
            .map(|items| MalObject::Vector(MalVector::new(items))),
        MalObject::Hashmap(MalHashmap { items, .. }) => items
            .iter()
            .try_fold(HashMap::new(), |mut acc, entry| {
                let (key, value) = (entry.0.clone(), entry.1.clone());
                let value = eval(&value, env)?;
                acc.insert(key, value);
                Ok(acc)
            })
            .map(|items| MalObject::Hashmap(MalHashmap::new(items))),
        MalObject::List(MalList { items, .. }) => match items.as_slice() {
            [MalObject::Symbol(MalSymbol { name, .. }), ops @ ..] => {
                let op = env.get(&name.clone().to_string());
                if let Some(op) = op {
                    let ops = ops.iter().try_fold(Vec::new(), |mut acc, arg| {
                        let item = eval(arg, env)?;
                        acc.push(item);
                        Ok(acc)
                    })?;
                    op.call(&ops)
                } else {
                    Err(MalError::RuntimeError(format!("Error: '{name}' not found")))
                }
            }
            _ => Ok(ast.clone()),
        },
        _ => Ok(ast.clone()),
    }
}

fn print(ast: &MalObject) -> String {
    println!("{:#?}", ast);
    pr_str(ast, true)
}

fn rep(input: &str) -> Result<String, MalError> {
    let env = repl_env();
    Ok(print(&eval(&read(input)?, &env)?))
}

fn main() {
    loop {
        let input = rl::readline("user> ").unwrap();
        let result = rep(&input);
        match result {
            Ok(output) => println!("{}", output),
            Err(MalError::ParseError(message) | MalError::RuntimeError(message)) => {
                println!("{}", message)
            }
            _ => {}
        }
    }
}

File: impls/ocaml541/bin/step2_eval.ml
open Ocaml541.Rl
open Ocaml541.Reader
open Ocaml541.Printer
open Ocaml541.Types
open Result.Syntax
module StringMap = Map.Make (String)

let traverse f lst =
  List.fold_left
    (fun acc x ->
      let* acc = acc in
      let* v = f x in
      Ok (v :: acc))
    (Ok []) lst
  |> Result.map List.rev

let read str = read_str str

let apply f args =
  match f with
  | Func (f, _) -> f args
  | _ -> error "attempt to call non-function"

let rec eval ast env =
  let eval' x = eval x env in
  match ast with
  | Sym s -> (
      match StringMap.find_opt s env with
      | Some r -> Ok r
      | None -> error (Format.asprintf "'%s' not found" s))
  | Vector (v, _) ->
      let* lst = traverse eval' v in
      Ok (vector lst)
  | Hash (hm, _) ->
      let entries = Hashtbl.to_seq hm |> List.of_seq in
      let* evalueted =
        traverse
          (fun (k, v) ->
            let* v' = eval' v in
            Ok (k, v'))
          entries
      in
      let new_hm = Hashtbl.create (List.length evalueted) in
      List.iter (fun (k, v) -> Hashtbl.add new_hm k v) evalueted;
      Ok (Hash (new_hm, Nil))
  | List (l, _) ->
      if List.length l = 0 then Ok ast
      else
        let a0 = List.hd l in
        let* f = eval a0 env in
        let* args = traverse eval' (List.tl l) in
        apply f args
  | _ -> Ok ast

let print ast = pr_str ast true

let rep str env =
  let* ast = read str in
  let* exp = eval ast env in
  Ok (print exp)

let int_op op a =
  match a with
  | [ Int a0; Int a1 ] -> Ok (Int (op a0 a1))
  | _ -> error "invalid int_op args"

let () =
  let env =
    StringMap.(
      empty
      |> add "+" (func (fun a -> int_op ( + ) a))
      |> add "-" (func (fun a -> int_op ( - ) a))
      |> add "*" (func (fun a -> int_op ( * ) a))
      |> add "/" (func (fun a -> int_op ( / ) a)))
  in
  let rec loop () =
    let result = readline ~prompt:"user> " () in
    match result with
    | Some line ->
        (match rep line env with
        | Ok out -> print_endline out
        | Error err ->
            let err_msg = Format.asprintf "Error: %s" (pr_str err true) in
            print_endline err_msg);
        loop ()
    | None -> loop ()
  in
  loop ()
